metricflow/dataflow/builder/dataflow_plan_builder.py

```diff
@@ -897,6 +886,19 @@ def _find_dataflow_recipe(
                 f"nodes for the right side of the join:\n"
                 f"{mf_pformat(candidate_nodes_for_right_side_of_join)}"
             )
+
+        # If there are MetricGroupBys in the requested linkable specs, build source nodes to satisfy them.
+        # We do this at query time instead of during usual source node generation because the number of potential
+        # MetricGroupBy source nodes could be extremely large (and potentially slow).
+        logger.info(f"Building source nodes for group by metrics: {linkable_spec_set.group_by_metric_specs}")
+        candidate_nodes_for_right_side_of_join += [
```

Comment:
Interesting. I think this is still fine for predicate pushdown, since they aren't really source nodes in the original sense, so I probably need to handle them separately anyway.